pkg/kubectl/cmd/set/set_port.go

```diff
@@ -0,0 +1,280 @@
+	o.Record = cmdutil.GetRecordFlag(cmd)
+	o.ChangeCause = f.Command(cmd, false)
+	o.PrintObject = f.PrintObject
+	o.Local = cmdutil.GetFlagBool(cmd, "local")
+	o.DryRun = cmdutil.GetDryRunFlag(cmd)
+	o.Output = cmdutil.GetFlagString(cmd, "output")
+	o.port = cmdutil.GetFlagInt(cmd, "port")
+	o.protocol = strings.ToUpper(cmdutil.GetFlagString(cmd, "protocol"))
+	o.containerSelector = cmdutil.GetFlagString(cmd, "containers")
+	o.overwrite = cmdutil.GetFlagBool(cmd, "overwrite")
+	o.Cmd = cmd
+
+	cmdNamespace, enforceNamespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+
+	builder := f.NewBuilder(!o.Local).
+		ContinueOnError().
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(enforceNamespace, &o.FilenameOptions).
+		Flatten()
+	if !o.Local {
+		builder = builder.
+			SelectorParam(o.Selector).
+			ResourceTypeOrNameArgs(o.All, args...).
+			Latest()
+	}
+	o.Infos, err = builder.Do().Infos()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *PortOptions) Validate() error {
+	errors := []error{}
+	if o.protocol != "TCP" && o.protocol != "UDP" {
+		errors = append(errors, fmt.Errorf("protocol must tcp or udp"))
+	}
+	if o.port <= 0 {
+		errors = append(errors, fmt.Errorf("must special port and port must greater than 0"))
+	}
+	return utilerrors.NewAggregate(errors)
+}
+
+func (o *PortOptions) Run() error {
+	allErrs := []error{}
+
+	patches := CalculatePatches(o.Infos, o.Encoder, func(info *resource.Info) ([]byte, error) {
+		transformed := false
+		_, err := o.UpdatePodSpecForObject(info.Object, func(spec *api.PodSpec) error {
+			containers, _ := selectContainers(spec.Containers, o.containerSelector)
+			if len(containers) == 0 {
+				fmt.Println(info.Mapping.Resource, info.Name, o.containerSelector)
+				if _, err := fmt.Fprintf(o.Err, "warning: %s/%s does not have any containers matching %q\n", info.Mapping.Resource, info.Name, o.containerSelector); err != nil {
+					return err
+				}
+				return nil
+			}
+
+			ports, err := o.portsForObject(info.Object)
+			if err != nil {
+				return err
+			}
+			if !o.overwrite {
+				for _, port := range ports {
+					if port == strconv.Itoa(o.port) {
+						return fmt.Errorf("'%s' already has a port %v, and --overwrite is false", c.Name, port)
+					}
+				}
+			}
+			for _, c := range containers {
+				for _, port := range c.Ports {
+					if port.ContainerPort != int32(o.port) || string(port.Protocol) != o.protocol {
+						transformed = true
+					}
+				}
+				containerPort := api.ContainerPort{
+					ContainerPort: int32(o.port),
+					Protocol:      api.Protocol(o.protocol),
+				}
+				c.Ports = append(c.Ports, containerPort)
+			}
+			if err != nil {
+				return err
+			}
+			return nil
+		})
+		if !transformed {
+			if _, err := fmt.Fprintln(o.Out, "no resources changed"); err != nil {
+				return nil, nil
+			}
+			return nil, err
+		}
+		if transformed && err == nil {
+			return runtime.Encode(o.Encoder, info.Object)
```

Comment:
rework this with the change from https://github.com/kubernetes/kubernetes/pull/53158/files#diff-466503ed5c81843c5ba705bc4c6fd4dc

Reply:
as is, this would fail intermittently on DaemonSet and ReplicaSet objects

Reply:
@liggitt nice, thank you very much.